### DISCUSSION TITLE
WASM GC should still have optional support for linear memory access

### DISCUSSION THREAD

**Discussion by lax1dude** [konsoletyper/teavm#966] [Ideas]
The more I think about it, there’s really not many practical uses for WASM that don’t also somehow involve rendering to a canvas. Using WASM in place of JavaScript for some responsive HTML based application is basically just a dumb gimmick, there’s no reason you’d actually want to use it instead of regular JavaScript for those kinds of applications besides to obfuscate the code more than you can achieve just using closure compiler, or of course to use the SIMD instructions to quietly mine cryptocurrency.

After looking into the WASM spec though, it’s clear that WASM GC in its current state isn’t gonna do much to help any sort of application that uses WebGL to render to a canvas. You can create byte arrays, but there’s no way to copy data between them besides using a `for` loop, and if you want to actually pass one to WebGL you gotta copy it one byte at a time into a typed array using an imported function, which is honestly embarrassing.

The easiest way to at least escape the overhead of calling an imported JavaScript once per byte of data you’re streaming to the GPU would be to still allow the WASM GC backend to import a `WebAssembly.Memory` object, with some simple malloc/free style allocator and the TeaVM `Address` class to actually load/store values to it. Allocations could then be wrapped with an NIO direct `ByteBuffer`, which is basically already how efficiently passing data to OpenGL via JNI on a regular JVM is done, so existing multi-platform TeaVM programs that render graphics using OpenGL/WebGL already probably store data meant to be passed to the GPU inside of a `ByteBuffer` instead of primitive arrays.

Of course we can also just wait around for WASM GC to get a copy instruction that can efficiently transfer data from a WASM GC array into a JavaScript typed array. However, considering it took 10 years for them to get around to just adding a copy instruction for conventional WASM to copy data efficiently in a memory, I wouldn’t get my hopes up that we’re even gonna see instructions for copying data efficiently between two WASM GC arrays soon, much less than efficiently copying data from a WASM GC array into a JavaScript typed array.

So the quickest way to remedy this using the functionality currently available would just be to allow WASM GC to have linear memory access, that `ByteBuffer` objects can use to store data that must be able to efficiently be transferred from WASM to WebGL without the overhead of calling an imported function once per byte that needs to be copied.

I don’t see any reason this would be hard to implement through recycling a few features from TeaVM’s existing legacy WASM backend.

**Comment by lax1dude** [comment]
@konsoletyper I'm willing to implement this ASAP if you don't currently think it's a priority, I'll do my best but of course it might take a little while since none of the WASM backend is documented and there aren't many comments in the source code of TeaVM either. If you plan to just wait for a more efficient method to become available for copying the contents of a WASM GC array to a JavaScript typed array then I will definitely implement support for linear memory in the WASM GC backend since I don't see why we must copy data at all when linear memory access for ByteBuffers would make it so we don't need to.

**Comment by lax1dude** [comment]
Alright now that I'm not at work trying to read and write all this on my phone, I see you have plans for a JSO extension where NIO buffers are backed by ArrayBuffers. In my app I created my own lightweight buffer classes to replace NIO a long time ago specifically for this purpose because the existing NIO buffers had too many features that didn't make sense on the web. In my opinion an optimal implementation of "direct buffers" in the WASM backend would be to make them slices of a conventional `WebAssembly.Memory` object, so that the conventional WASM memory load/store instructions can be used for accessing them instead of having to call imported functions from JavaScript land to manipulate them. This would be WAY faster than probably even what future additions to the WASM GC spec could do for us if they add dedicated instructions for manipulating the contents of a plain ArrayBuffer. We can completely eliminate the overhead (for the time being) of having to import functions to access an ArrayBuffer if we just make direct buffers slices of a `WebAssembly.Memory` object instead. The only issue is it reduces the interoperability with JavaScript builds and requires the buffers to be explicitly free'd, however this is already how buffers in LWJGL works so I don't think its necessarily that big of a deal to require NIO buffers to (in this theoretical model) be explicitly free'd instead of garbage collected automatically.

**Reply by lax1dude** [reply]
@xpenatan Am I correct in assuming that libGDX already has everything in place to explicitly free NIO buffers when using LWJGL, and that functionality could easily be carried over to the web?

**Reply by xpenatan** [reply]
I'm not sure. Libgdx use allocateDirect for all buffers that is going to be used.  All libgddx opengl command needs a NIO buffer. For the web we need typearrays so we just copy NIO buffer created by teavm buffers to typearrays before sending to webgl. For javascript I did some emualtion tricks to obtain typearray from teavm buffers without copying but that does not work for webassembly.

**Reply by lax1dude** [reply]
Damn, so we would need finalizer support or create some system using weak references in that case for the regular JRE allocateDirect to work. In my app I went the route of LWJGL 3 where my custom direct buffers are allocated through a jemalloc based buffer allocator class instead of the JRE and not free’d automatically via the finalizer, and then have my allocator class’s free function be a NOP in JavaScript builds where they are backed by ArrayBuffers. All the related OpenGL resources already need to be free’d manually anyway so it’s not that much more of a burden to manually free buffers as well, and I know from the time I tried to do it when I was 12 years old that freeing my OpenGL objects via finalizer is dumb and causes lots of sneaky use-after-free issues without “pinning” certain objects, and freeing my buffers via finalizer would come with most of the same risks as well. I can’t imagine any reason though that having direct buffers backed by externref ArrayBuffers would give better performance than having direct buffers that exist as slices of a WebAssembly.Memory, so I don’t think we should completely discount the idea just because they don’t garbage collect perfectly. I’m probably just gonna fork TeaVM for me and the community around my project if Alexey insists on only backing them with ArrayBuffers and assume that WASM GC will eventually get magic instructions that can somehow manipulate the contents of an externref ArrayBuffer the same speed as using a WebAssembly.Memory sometime in the next decade (which is not happening, there’s no way it could ever come close to the same performance for operations other than bulk copy, even with dedicated instructions).

**Comment by konsoletyper** [comment]
> The more I think about it, there’s really not many practical uses for WASM that don’t also somehow involve rendering to a canvas. 

It's nothing more than your opinion

> Using WASM in place of JavaScript for some responsive HTML based application is basically just a dumb gimmick

First, it's opinionated. Second, there are use cases beyond both "responsive HTML based applications" and "involve rendering to a canvas"

> You can create byte arrays, but there’s no way to copy data between them besides using a for loop, and if you want to actually pass one to WebGL you gotta copy it one byte at a time into a typed array using an imported function, which is honestly embarrassing.

Yes, it's a shame on WebAssembly committee.

BUT! Copying data between main memory and GPU memory is anyway is a slow operation and should be performed as rare as possible. If you send gigabytes of data to GPU every frame, you'll get troubles *without* WebAssembly

> Of course we can also just wait around for WASM GC to get a copy instruction that can efficiently transfer data from a WASM GC array into a JavaScript typed array

No, copy instruction is a bad solution. A good solution is either a way to represent Wasm GC array as `ArrayBuffer` in JS *or* to allow to pass Wasm GC arrays to WebGL/WebGPU.

> However, considering it took 10 years for them to get around to just adding a copy instruction for conventional WASM to copy data efficiently in a memory

I think it was like one or two years.

> So the quickest way to remedy this using the functionality currently available would just be to allow WASM GC to have linear memory access, that ByteBuffer objects can use to store data that must be able to efficiently be transferred from WASM to WebGL without the overhead of calling an imported function once per byte that needs to be copied.

Sure, that's *exactly* what I was going to do, which I made perfectly clear in my post.

> I don’t see any reason this would be hard to implement through recycling a few features from TeaVM’s existing legacy WASM backend.

There's nothing difficult, it just requires some time, which right now I don't have.

> I'm willing to implement this ASAP

No, please, just don't

> if you don't currently think it's a priority

I *do* think it's a priority for TeaVM, but for my personal life TeaVM is not priority right now. I must pass B1 Deutsch exam to be able to stay in Germany without visiting every couple years to *Ausländerbehörde*, as well as doing some paperwork. 

> In my opinion an optimal implementation of "direct buffers" in the WASM backend would be to make them slices of a conventional WebAssembly.Memory object, so that the conventional WASM memory load/store instructions can be used for accessing them instead of having to call imported functions from JavaScript land to manipulate them

Did you read my post carefuly? I never claimed I'm going to use any imported functions.

**Reply by lax1dude** [reply]
Okay thanks for reading my huge wall of text. I've had nothing better to do for the past week so I've been working on trying to implement a memory allocator myself on a branch. Its still a work in progress and I need to test it but I think its a step in the right direction. I put as many comments as possible on the actual memory allocator class so you can tell how it works.

[https://github.com/konsoletyper/teavm/compare/master...lax1dude:teavm:841e09cbd5a880fd1a0a3287913f2de056e082c4](https://github.com/konsoletyper/teavm/compare/master...lax1dude:teavm:841e09cbd5a880fd1a0a3287913f2de056e082c4)

**Reply by lax1dude** [reply]
Is any of this usable?

**Reply by lax1dude** [reply]
@konsoletyper Okay I went back and read every comment on your WASM announcement and I found your reply to the libGDX guy where you said "I'm going to slightly different thing: implement Java NIO buffers over Wasm heap". I was not aware of this, I'm sorry I didn't read every last comment on the announcement. Please check out my branch though, specifically LaxMalloc.java, I basically recreated the fundementals of the emscripten memory allocator (It has O(1) complexity for most allocations) using TeaVM's Address class for heap access, and I think you might find it useful. Again I am sorry I didn't see your comment but to be fair it wasn't part of the main announcement so I think you can understand how I overlooked it. I am only trying to help, I tried my very best to implement it in a way you would agree with.

[`core/src/main/java/org/teavm/runtime/LaxMalloc.java`](https://github.com/lax1dude/teavm/blob/wasm-gc-linear-memory/core/src/main/java/org/teavm/runtime/LaxMalloc.java)

**Reply by konsoletyper** [reply]
@lax1dude sorry, but I won't review this or approve any PR regarding this feature

**Reply by lax1dude** [reply]
> @lax1dude sorry, but I won't review this or approve any PR regarding this feature

@konsoletyper I never said pull request, I suggested you check out LaxMalloc.java where I’ve implemented the emscripten memory allocator, since it might be helpful. I can tell you’ve developed this perception of me that I’m some impotent fool, I don’t know why because you know literally nothing about me, and I don’t know why you’re so always eager to always dismiss me at every corner either. It’s really demoralizing and I don’t understand why would want to treat your community that way.
